import { Repository } from "typeorm";
import { User } from "../entity/User";
import { UpdateUserData, UserData } from "../types";
import createHttpError from "http-errors";
import bcrypt from "bcrypt";

export class UserService {
    constructor(private userRepository: Repository<User>) {}
    async create({ firstName, lastName, email, password, role }: UserData) {
        const user = await this.userRepository.findOne({
            where: { email: email },
        });
        if (user) {
            const error = createHttpError(400, "Email is already exist");
            throw error;
        }
        // Hash Password
        const saltRounds = 10;
        const hashedPassword = await bcrypt.hash(password, saltRounds);
        try {
            const user = await this.userRepository.save({
                firstName,
                lastName,
                email,
                password: hashedPassword,
                role,
            });
            return user;
        } catch (err) {
            const error = createHttpError(
                500,
                "faild to store data in database",
            );
            throw error;
        }
    }

    async findByEmail(email: string) {
        return await this.userRepository.findOne({
            where: {
                email,
            },
            select: [
                "id",
                "firstName",
                "lastName",
                "email",
                "role",
                "password",
            ],
        });
    }

    async findById(id: number) {
        return await this.userRepository.findOne({
            where: {
                id,
            },
        });
    }

    async getAll() {
        return await this.userRepository.find();
    }

    async update(
        { firstName, lastName, role }: UpdateUserData,
        userId: number,
    ) {
        try {
            const user = await this.userRepository.update(userId, {
                firstName,
                lastName,
                role,
            });
            return user;
        } catch (err) {
            const error = createHttpError(
                500,
                "Failed to update the user in the database",
            );
            throw error;
        }
    }
    async delete(userId: number) {
        return await this.userRepository.delete(userId);
    }
}
